File: construction/construction/report/wbs_tree/wbs_tree.js
frappe.query_reports["WBS Tree"] = {
    filters: [
        {
            fieldname: "cost_code",
            label: __("Cost Code"),
            fieldtype: "Data",
        },
        {
            fieldname: "cost_center_code",
            label: __("Cost Center Code"),
            fieldtype: "Data",
        },
        {
            fieldname: "project",
            label: __("Project"),
            fieldtype: "Link",
            options: "Project"
        },
        {
            fieldname: "item",
            label: __("Item"),
            fieldtype: "Link",
            options: "Item"
        }
    ],

    onload: function(report) {
        this._report = report;
        this._initializeReport();
    },

    _initializeReport: function() {
        this._addCustomStyles();
        this._setupEventHandlers();
    },

    _addCustomStyles: function() {
        const styles = document.createElement('style');
        styles.textContent = `
            .wbs-buttons {
                white-space: nowrap;
                display: inline-flex;
            }
            .wbs-buttons .btn {
                margin-right: 2px;
            }
            .wbs-buttons .btn-xs {
                padding: 1px 5px;
                font-size: 12px;
                line-height: 1.5;
                border-radius: 3px;
            }
        `;
        document.head.appendChild(styles);
    },

    _setupEventHandlers: function() {
        // Remove any existing event handlers
        $(document).off('click.wbsReport', '.add-wbs-btn, .goto-wbs-btn, .edit-ref-btn, .create-stock-entry-btn');
        
        // Go to WBS button handler
        $(document).on('click.wbsReport', '.goto-wbs-btn', (e) => {
            e.preventDefault();
            const $button = $(e.currentTarget);
            const wbsName = $button.attr('data-wbs-name');
            
            if (wbsName) {
                frappe.set_route('Form', 'WBS Item', wbsName);
            } else {
                frappe.msgprint(__('Unable to find WBS Item reference. Please refresh the report.'));
            }
        });

        // Add WBS button handler
        $(document).on('click.wbsReport', '.add-wbs-btn', (e) => {
            e.preventDefault();
            const $button = $(e.currentTarget);
            const wbsName = $button.attr('data-wbs-name');
            
            if (wbsName) {
                this._showAddWBSDialog({ name: wbsName });
            }
        });

        // Edit reference button handler
        $(document).on('click.wbsReport', '.edit-ref-btn', (e) => {
            e.preventDefault();
            const $button = $(e.currentTarget);
            const wbsName = $button.attr('data-wbs-name');
            
            if (wbsName) {
                this._showEditRefDialog({ name: wbsName });
            }
        });

        // Create Stock Entry button handler
        $(document).on('click.wbsReport', '.create-stock-entry-btn', (e) => {
            e.preventDefault();
            const $button = $(e.currentTarget);
            const wbsName = $button.attr('data-wbs-name');
            
            if (wbsName) {
                this._showCreateStockEntryDialog({ name: wbsName });
            }
        });
    },

    _showAddWBSDialog: function(wbsData) {
        const dialog = new frappe.ui.Dialog({
            title: __('Add WBS Item'),
            fields: [
                {
                    label: __('Parent WBS Item'),
                    fieldname: 'parent_wbs_item',
                    fieldtype: 'Data',
                    read_only: 1,
                    default: wbsData.name
                },
                {
                    label: __('Item'),
                    fieldname: 'item',
                    fieldtype: 'Link',
                    options: 'Item'
                },
                {
                    label: __('Item Group'),
                    fieldname: 'item_group',
                    fieldtype: 'Link',
                    options: 'Item Group'
                },
                {
                    label: __('BOQ Qty'),
                    fieldname: 'qty',
                    fieldtype: 'Float',
                    default: 0
                },
                {
                    label: __('Budget'),
                    fieldname: 'budget',
                    fieldtype: 'Currency',
                    default: 0
                },
                {
                    label: __('Budget Rate'),
                    fieldname: 'unit_rate',
                    fieldtype: 'Currency',
                    default: 0
                },
            ],
            primary_action_label: __('Create'),
            primary_action(values) {
                frappe.call({
                    method: 'frappe.client.insert',
                    args: {
                        doc: {
                            doctype: 'WBS item',
                            parent_wbs_item: values.parent_wbs_item,
                            item: values.item,
                            item_group: values.item_group,
                            unit_rate: values.unit_rate,
                            qty: values.qty,
                            budget: values.budget,
                        }
                    },
                    callback: function(r) {
                        if (!r.exc) {
                            frappe.show_alert({
                                message: __('WBS Item created successfully'),
                                indicator: 'green'
                            });
                            dialog.hide();
                            frappe.query_report.refresh();
                        }
                    }
                });
            }
        });
        dialog.show();
    },

    _showEditRefDialog: function(wbsData) {
        // First fetch the document data
        frappe.call({
            method: 'frappe.client.get',
            args: {
                doctype: 'WBS item',
                name: wbsData.name
            },
            callback: (response) => {
                if (response.message) {
                    const doc = response.message;
                    const dialog = new frappe.ui.Dialog({
                        title: __('Edit WBS Item'),
                        fields: [
                            {
                                label: __('Serial'),
                                fieldname: 'serial',
                                fieldtype: 'Data',
                                default: doc.serial
                            },
                            {
                                label: __('Serial No'),
                                fieldname: 'serial_no',
                                fieldtype: 'Data',
                                default: doc.serial_no
                            },
                            {
                                label: __('Short Description'),
                                fieldname: 'short_description',
                                fieldtype: 'Data',
                                default: doc.short_description
                            },
                            {
                                label: __('Item'),
                                fieldname: 'item',
                                fieldtype: 'Link',
                                options: 'Item',
                                default: doc.item
                            },
                            {
                                label: __('Item Group'),
                                fieldname: 'item_group',
                                fieldtype: 'Link',
                                options: 'Item Group',
                                default: doc.item_group
                            },
                            {
                                label: __('UOM'),
                                fieldname: 'uom',
                                fieldtype: 'Link',
                                options: 'UOM',
                                default: doc.uom
                            },
                            {
                                label: __('BOQ Qty'),
                                fieldname: 'qty',
                                fieldtype: 'Float',
                                default: doc.qty
                            },
                            {
                                label: __('Consumed Quantity'),
                                fieldname: 'consumed_quantity',
                                fieldtype: 'Float',
                                default: doc.consumed_quantity
                            },
                            {
                                label: __('Material Rate'),
                                fieldname: 'unit_cost',
                                fieldtype: 'Currency',
                                default: doc.unit_cost
                            },
                            {
                                label: __('Budget'),
                                fieldname: 'budget',
                                fieldtype: 'Currency',
                                default: doc.budget
                            },
                            {
                                label: __('PR Reserved Qty'),
                                fieldname: 'pr__reserved_qty',
                                fieldtype: 'Float',
                                default: doc.pr__reserved_qty
                            },
                            {
                                label: __('PO Reserved Qty'),
                                fieldname: 'po_reserved_qty',
                                fieldtype: 'Float',
                                default: doc.po_reserved_qty
                            },
                            {
                                label: __('Available Qty'),
                                fieldname: 'available_qty',
                                fieldtype: 'Float',
                                default: doc.available_qty
                            },
                            {
                                label: __('PO Reserved Amount'),
                                fieldname: 'total',
                                fieldtype: 'Currency',
                                default: doc.total
                            }
                        ],
                        primary_action_label: __('Update'),
                        primary_action(values) {
                            // Create an object with only the changed values
                            const updates = {};
                            for (const [key, value] of Object.entries(values)) {
                                if (value !== doc[key]) {
                                    updates[key] = value;
                                }
                            }

                            if (Object.keys(updates).length === 0) {
                                dialog.hide();
                                return;
                            }

                            frappe.call({
                                method: 'frappe.client.set_value',
                                args: {
                                    doctype: 'WBS item',
                                    name: wbsData.name,
                                    fieldname: updates
                                },
                                callback: function(r) {
                                    if (!r.exc) {
                                        frappe.show_alert({
                                            message: __('WBS Item updated successfully'),
                                            indicator: 'green'
                                        });
                                        dialog.hide();
                                        frappe.query_report.refresh();
                                    }
                                }
                            });
                        }
                    });
                    dialog.show();
                }
            }
        });
    },
    
    _showCreateStockEntryDialog: function(wbsData) {
        // First fetch the WBS item details
        frappe.call({
            method: 'frappe.client.get',
            args: {
                doctype: 'WBS item',
                name: wbsData.name
            },
            callback: (response) => {
                if (response.message) {
                    const wbsDoc = response.message;
                    
                    // Create the dialog for stock entry
                    const dialog = new frappe.ui.Dialog({
                        title: __('Create Purchase Request'),
                        fields: [
                            {
                                label: __('WBS Item'),
                                fieldname: 'wbs_item',
                                fieldtype: 'Link',
                                options: 'WBS Item',
                                default: wbsData.name,
                                read_only: 1
                            },
                            {
                                label: __('Item'),
                                fieldname: 'item',
                                fieldtype: 'Link',
                                options: 'Item',
                                default: wbsDoc.item,
                                read_only: 1
                            },
                            {
                                label: __('Available Quantity'),
                                fieldname: 'available_qty',
                                fieldtype: 'Float',
                                default: wbsDoc.available_qty,
                                read_only: 1
                            },
                            {
                                label: __('Quantity'),
                                fieldname: 'qty',
                                fieldtype: 'Float',
                                reqd: 1
                            },
                            {
                                label: __('Source Warehouse'),
                                fieldname: 'source_warehouse',
                                fieldtype: 'Link',
                                options: 'Warehouse',
                                reqd: 1
                            }
                        ],
                        primary_action_label: __('Create'),
                        primary_action(values) {
                            // Validate quantity
                            if (values.qty <= 0) {
                                frappe.msgprint({
                                    title: __('Error'),
                                    indicator: 'red',
                                    message: __('Quantity must be greater than zero')
                                });
                                return;
                            }
                            
                            // Validate against available quantity
                            if (values.qty > wbsDoc.available_qty) {
                                frappe.msgprint({
                                    title: __('Error'),
                                    indicator: 'red',
                                    message: __('Insufficient quantity available. Requested: {0}, Available: {1}', 
                                        [values.qty, wbsDoc.available_qty])
                                });
                                return;
                            }
                            
                            // Create Stock Entry
                            frappe.call({
                                method: "frappe.client.insert",
                                args: {
                                    doc: {
                                        doctype: "Material Request",
                                        material_request_type: "Purchase",
                                        project: wbsDoc.project,
                                        items: [{
                                            item_code: wbsDoc.item,
                                            qty: values.qty,
                                            schedule_date: frappe.datetime.now_date(),
                                            warehouse: values.source_warehouse,
                                            custom_wbs: wbsData.name,
                                            custom_cost_code: wbsDoc.cost_code,
                                            custom_combined_code: wbsDoc.combined_code,
                                            rate: wbsDoc.unit_rate
                                        }]
                                    }
                                },
                                callback: function(response) {
                                    if (response.message) {
                                        frappe.show_alert({
                                            message: __('Purchase Request created successfully'),
                                            indicator: 'green'
                                        });
                                        dialog.hide();
                                        frappe.set_route("Form", "Material Request", response.message.name);
                                    }
                                }
                            });
                        }
                    });
                    
                    dialog.show();
                } else {
                    frappe.msgprint(__('Unable to fetch WBS Item details. Please try again.'));
                }
            }
        });
    },

    _generateButtons: function(data) {
        // Safety check for data
        if (!data) {
            console.warn('No data provided to _generateButtons');
            return '';
        }
    
        // Get the WBS name from the appropriate field
        const wbsName = data.name || data.wbs_item || data.id;
        
        if (!wbsName) {
            console.warn('No WBS name found in data:', data);
            return '';
        }
    
        // "Add" button should always appear
        let buttons = `
            <div class="wbs-buttons">
                <button type="button" class="btn btn-xs btn-info goto-wbs-btn" 
                    data-wbs-name="${frappe.utils.escape_html(wbsName)}"
                    title="Go to WBS Item">
                    <i class="fa fa-external-link"></i>
                </button>`;

        if (data.level < 5){
            buttons += `<button type="button" class="btn btn-xs btn-success add-wbs-btn" 
                    data-wbs-name="${frappe.utils.escape_html(wbsName)}"
                    title="Add WBS Item">
                    <i class="fa fa-plus"></i>
                </button>`
        }
        if (data.level == 5){
            buttons += `
                <button type="button" class="btn btn-xs btn-warning edit-ref-btn" 
                    data-wbs-name="${frappe.utils.escape_html(wbsName)}"
                    title="Edit Reference">
                    <i class="fa fa-pencil"></i>
                </button>
                <button type="button" class="btn btn-xs btn-dark create-stock-entry-btn" 
                    data-wbs-name="${frappe.utils.escape_html(wbsName)}"
                    title="Create Stock Entry">
                    Create Purchase Request
                </button>`
        }
    
        buttons += `</div>`;
        return buttons;
    },
    
    formatter: function(value, row, column, data, default_formatter) {
        if (column.fieldname === "actions" && data) {
            // If data is a string, try to parse it
            if (typeof data === 'string') {
                try {
                    data = JSON.parse(data);
                } catch (e) {
                    console.error('Error parsing data:', e);
                }
            }
            return this._generateButtons(data);
        }
        return default_formatter(value, row, column, data);
    }
};

File: construction/construction/report/wbs_tree/wbs_tree.py
from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import cstr

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    
    # Debug information
    if not data:
        frappe.msgprint("No data found after processing. Raw query returned data but post-processing resulted in empty dataset.")
    
    return columns, data

def get_columns():
    return [
        {
            "fieldname": "cost_code",
            "label": _("Cost Code"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "cost_center_code",
            "label": _("Financial Code"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "item_group",
            "label": _("Item Group"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "combined_code",
            "label": _("Combined Code"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "short_description",
            "label": _("Short Description"),
            "fieldtype": "Data",
            "width": 250
        },
        {
            "fieldname": "item",
            "label": _("Item"),
            "fieldtype": "Data",
            "width": 100
        },
         {
            "fieldname": "uom",
            "label": _("UOM"),
            "fieldtype": "Data",
            "width": 100
        },
         {
            "fieldname": "qty",
            "label": _("BOQ Qty"),
            "fieldtype": "Float",
            "width": 100
        },
         {
            "fieldname": "available_qty",
            "label": _("Available Qty"),
            "fieldtype": "Float",
            "width": 100
        },
         {
            "fieldname": "resource_qty",
            "label": _("Resource Qty"),
            "fieldtype": "Float",
            "width": 100
        },
        {
            "fieldname": "waste",
            "label": _("Waste"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "unit_cost",
            "label": _("Material Rate"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "unit_rate",
            "label": _("Budget Rate"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "budget",
            "label": _("Budget"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "total",
            "label": _("PO Reserved Amount"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "actions",
            "label": _(""),
            "fieldtype": "Data",
            "width": 250
        },
    ]

def get_wbs_items(filters):
    conditions = []
    
    if filters and filters.get('parent_wbs_item'):
        conditions.append(f"parent_wbs_item = '{filters.get('parent_wbs_item')}'")
    
    if filters and filters.get('cost_code'):
        conditions.append(f"cost_code LIKE '%{filters.get('cost_code')}%'")
    
    if filters and filters.get('cost_center_code'):
        conditions.append(f"cost_center_code LIKE '%{filters.get('cost_center_code')}%'")
    
    if filters and filters.get('project'):
        conditions.append(f"project = '{filters.get('project')}'")
    
    if filters and filters.get('item'):
        conditions.append(f"item = '{filters.get('item')}'")
    
    where_clause = " AND ".join(conditions) if conditions else "1=1"
    
    return frappe.db.sql(f"""
        SELECT 
            name,
            cost_code,
            cost_center_code,
            unit_rate,
            resource_qty,
            combined_code,
            serial,
            serial_no,
            qty,
            short_description,
            item_group,
            consumed_quantity,
            po_reserved_qty,
            available_qty,
            waste,
            pr__reserved_qty,
            budget,
            total,
            item,
            unit_cost,
            uom,
            level,
            parent_wbs_item,
            is_group
        FROM
            `tabWBS item`
        WHERE 
            {where_clause}
        ORDER BY serial, name
    """, as_dict=True)

def get_data(filters):
    wbs_items = get_wbs_items(filters)
    
    if not wbs_items:
        frappe.msgprint("No items returned from database query")
        return []
    
    # When dealing with filters, we need a different approach to ensure proper tree view
    # We'll fetch all parents of filtered items to maintain the hierarchy
    
    # Collect all filtered items
    filtered_items = {item.name: item for item in wbs_items}
    
    # Add parent-child relationships for the filtered items
    parent_child_map = {}
    all_parents = set()
    
    for item in wbs_items:
        if item.parent_wbs_item:
            parent_child_map.setdefault(item.parent_wbs_item, []).append(item.name)
            all_parents.add(item.parent_wbs_item)
    
    # Find parent items that aren't in our filtered result but need to be included
    missing_parents = all_parents - set(filtered_items.keys())
    
    # If we have missing parents, fetch them from the database
    if missing_parents:
        parent_items = frappe.db.sql(f"""
            SELECT 
                name,
                cost_code,
                cost_center_code,
                unit_rate,
                resource_qty,
                combined_code,
                serial,
                serial_no,
                qty,
                short_description,
                item_group,
                consumed_quantity,
                po_reserved_qty,
                available_qty,
                waste,
                pr__reserved_qty,
                budget,
                total,
                item,
                unit_cost,
                uom,
                level,
                parent_wbs_item,
                is_group
            FROM
                `tabWBS item`
            WHERE 
                name IN ({', '.join(["'" + p + "'" for p in missing_parents])})
        """, as_dict=True)
        
        # Add these parents to our filtered items
        for item in parent_items:
            filtered_items[item.name] = item
            if item.parent_wbs_item:
                parent_child_map.setdefault(item.parent_wbs_item, []).append(item.name)
                all_parents.add(item.parent_wbs_item)
    
    # Reconstruct the tree from top to bottom
    # First find all root items (items with no parent or parent not in our dataset)
    root_items = [name for name, item in filtered_items.items() 
                if not item.parent_wbs_item or item.parent_wbs_item not in filtered_items]
    
    # Sort root items by serial
    root_items.sort(key=lambda x: (filtered_items[x].serial or "", x))
    
    # Process the tree
    data = []
    
    def process_tree(name, level=0):
        item = filtered_items[name]
        
        data.append({
            "name": item.name,
            "unit_rate": item.unit_rate,
            "cost_code": item.cost_code,
            "resource_qty": item.resource_qty,
            "waste": item.waste,
            "combined_code": item.combined_code,
            "cost_center_code": item.cost_center_code,
            "item": item.item,
            "level": item.level,
            "unit_cost": item.unit_cost,
            "serial_no": item.serial_no,
            "po_reserved_qty": item.po_reserved_qty,
            "available_qty": item.available_qty,
            "pr__reserved_qty": item.pr__reserved_qty,
            "short_description": item.short_description,
            "item_group": item.item_group,
            "consumed_quantity": item.consumed_quantity,
            "budget": item.budget,
            "total": item.total,
            "uom": item.uom,
            "parent_wbs_item": item.parent_wbs_item,
            "indent": level,
            "qty": item.qty,
            "is_group": item.is_group
        })
        
        # Process children if any
        children = parent_child_map.get(name, [])
        children.sort(key=lambda x: (filtered_items[x].serial or "", x))
        
        for child in children:
            process_tree(child, level + 1)
    
    # Process each root item
    for root in root_items:
        process_tree(root)
    
    # If no data generated through the tree, fall back to showing all filtered items
    if not data and filtered_items:
        for name, item in filtered_items.items():
            data.append({
                "name": item.name,
                "unit_rate": item.unit_rate,
                "cost_code": item.cost_code,
                "resource_qty": item.resource_qty,
                "waste": item.waste,
                "combined_code": item.combined_code,
                "cost_center_code": item.cost_center_code,
                "item": item.item,
                "unit_cost": item.unit_cost,
                "serial_no": item.serial_no,
                "po_reserved_qty": item.po_reserved_qty,
                "available_qty": item.available_qty,
                "pr__reserved_qty": item.pr__reserved_qty,
                "short_description": item.short_description,
                "item_group": item.item_group,
                "consumed_quantity": item.consumed_quantity,
                "budget": item.budget,
                "total": item.total,
                "uom": item.uom,
                "parent_wbs_item": item.parent_wbs_item,
                "indent": 0,
                "qty": item.qty,
                "is_group": item.is_group
            })
    
    return data